### PR TITLE
DEV-86 - translate coffeescript to typescript

### PR DESCRIPTION
**Pull request by DaleHerring**

### CODE REVIEW

**Review by bnlb — CHANGES_REQUESTED** [review]
Blockers are:

- Readme + examples
- Using `lib` instead of `build`.
- Updating tests to use a new account

File: src/HttpClient.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+class HttpClient {
+
+    private key: string;
+    private headers: Object;
+    private httpClient: hC.HttpClient;
+
+    public constructor(key: string) {
+        this.key = key;
+        this.headers = { 'Authorization': this.key, 'Content-Type': 'application/json' };
+        this.httpClient = new hC.HttpClient('');
```
[the remark this answers is not in the document]
Comment (DaleHerring) [reply src/HttpClient.ts:12]: The useragent info is supposed to go there, I just forgot to add it back in, fix inc.

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export { Algorithmia } from '../src/Algorithmia';
```
Comment (bnlb) [inline src/index.ts:1]:
```suggestion
export { Algorithmia as default } from './Algorithmia';
```

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+export { AlgorithmiaClient } from './AlgorithmiaClient';
+export { DataFile, DataDir, DataList } from '../src/Data';
```
Comment (bnlb) [inline src/index.ts:3]:
```suggestion
export { DataFile, DataDir, DataList } from './Data';
```

File: src/HttpClient.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  public constructor(key: string) {
+    this.key = key;
+    this.headers['Authorization'] = this.key;
+    this.httpClient = new TypedHttpClient(this.userAgent);
+  }
+
+  async get(path: string) {
+    return await this.httpClient.get(path, this.headers).then((x) => {
+      return x.readBody();
+    });
```
Comment (bnlb) [inline src/HttpClient.ts:24]: Let's update these functions to `await` responses instead of using promise chains. It's more intuitive for most people and more in line with current best practices.

```
async get(path: string) {
  const response = await this.httpClient.get(path, this.headers);

  return response.readBody();
}
```

File: src/Data.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  delete() {
+    return this.client.delete(this.path);
+  }
+
+  exists() {
+    return this.client.head(this.path);
+  }
+}
```
Comment (bnlb) [inline src/Data.ts:46]: The current Node client has [a lot more methods](https://github.com/algorithmiaio/algorithmia-nodejs/blob/master/lib/data.js#L33-L116) than are listed here. In order to keep our public API the same, they'll need to be added and match the same function signatures.

```
put(content, callback)

get(callback)

putString(content, callback)

putJson(content, callback)

getString(callback)

getJson(callback)

exists(callback)

delete(callback)
```

I like what you've done here returning promises directly, as that's more in line with what I'd expect as a JS dev, but I don't want to introduce an API change as a part of this PR. Our API should remain callback based for now. We can update to promises in a separate PR.

File: src/Data.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    return this.client.post(this.path, input, contentType);
+  }
+
+  exists() {
+    return this.client.head(this.path);
+  }
+
+  delete(force: boolean) {
+    return this.client.delete(`${this.path}?force=${force}`);
+  }
```
Comment (bnlb) [inline src/Data.ts:79]: We'll need all of the [current methods](https://github.com/algorithmiaio/algorithmia-nodejs/blob/master/lib/data.js#L118-L205) on the Directory instance as well:

```
create(callback)

file(filename)

putFile(filePath, callback)

iterator()

forEach(callback)

forEachFile(callback) 

forEachDir(callback)

exists(callback)

delete(force, callback)
```

File: src/AlgorithmExecutable.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    input: Object,
+    version?: string,
+    output = 'raw',
+    stdout = false,
+    timeout = 300
+  ) {
+    const contentTypeHelper: ContentTypeHelper = new ContentTypeHelper();
+    let contentType: string;
+
+    contentType = contentTypeHelper.contentTypeHelper(input);
```
Comment (bnlb) [inline src/AlgorithmExecutable.ts:23]: The benefit of changing the contentType helper to a function instead of a class is that it simplifies the cases where it's used:

```
const contentType = getContentType(input);
```

File: src/AlgorithmExecutable.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+          version +
+          '/?output=' +
+          output +
+          '&stdout=' +
+          stdout +
+          '&timeout=' +
+          timeout,
+        input,
+        contentType
+      );
```
Comment (bnlb) [inline src/AlgorithmExecutable.ts:50]: This is hard to follow. Can simplify a bit:

```
const pathname = version
  ? `${this.path}/${version}/`
  : `${this.path}/`;

const params = new URLSearchParams({
  output,
  stdout,
  timeout,
});

const fullPath = `${pathname}?${params.toString()}`;

return this.client.post(fullPath, input, contentType);
```

File: src/AlgorithmiaClient.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      this.defaultApiAddress + this.dataPrefix + path
+    );
+  }
+
+  /**
+   * Initialize a DataDirectory object from this client
+   * @param path to a data directory, e.g., data://.my/foo
+   * @return a DataDirectory client for the specified directory
+   */
+  dir(path: string): DataDir {
```
Comment (bnlb) [inline src/AlgorithmiaClient.ts:60]: nit: The return types on these functions can be inferred, so you can drop them if you'd prefer and TypeScript will still know what's happening: `dir(path: string) {`

File: README.md [hunk 10]
```diff
@@ -121,128 +87,61 @@ The Algorithmia client also provides a way to manage both Algorithmia hosted dat
 
 ### Create directories
 
-Create directories by instantiating a `Dir` object and calling `create()`:
+Create directories by instantiating a `DataDir` object and calling `post()`:
```
Comment (bnlb) [inline README.md:90]: Let's change this method name back to `create`. Even though a POST is what's happening under the hood, from a user's perspective it's more intuitive to create a directory.

File: package.json [hunk 11]
```diff
@@ -1,22 +1,51 @@
...
+    "@typescript-eslint/eslint-plugin": "^4.11.0",
+    "@typescript-eslint/parser": "^4.11.0",
+    "babel-jest": "^26.6.3",
+    "eslint": "^7.16.0",
+    "eslint-config-prettier": "^7.1.0",
+    "eslint-plugin-prettier": "^3.3.0",
+    "husky": "^4.3.6",
+    "install": "^0.13.0",
+    "jest": "^26.6.3",
+    "npm": "^6.14.9",
```
Comment (bnlb) [inline package.json:42]: Are we using `npm` anywhere in the repo? I've never seen it installed in any project I've worked on. Should be safe to remove.

File: package.json [hunk 12]
```diff
@@ -1,22 +1,51 @@
...
+    "@types/jest": "^26.0.15",
+    "@types/node": "^14.11.2",
+    "@typescript-eslint/eslint-plugin": "^4.11.0",
+    "@typescript-eslint/parser": "^4.11.0",
+    "babel-jest": "^26.6.3",
+    "eslint": "^7.16.0",
+    "eslint-config-prettier": "^7.1.0",
+    "eslint-plugin-prettier": "^3.3.0",
+    "husky": "^4.3.6",
+    "install": "^0.13.0",
```
Comment (bnlb) [inline package.json:40]: What's this? Should be safe to remove.

File: README.md [hunk 13]
```diff
@@ -37,69 +37,36 @@ The returned promise will be called with the response with the Algorithm complet
...
 ```javascript
-client.algo('util/whoopsWrongAlgo').pipe('Hello, world!')
-      .then(function(response) {
-          if(response.error) {
-            console.log("Error: " + response.error.message);
-          } {
-            console.log(response.get());
-          }
-      });
+client.algo('util/whoopsWrongAlgo').pipe('Hello, world!').then(x => { console.log(x) });
```
Comment (bnlb) [inline README.md:69]: In all of the cases where we're demonstrating a response to a user, let's await instead.

```
const response = await client.algo('util/whoopsWrongAlgo').pipe('Hello, world!');
```

File: test/Data.test.ts [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+      expect(await file.exists()).toBe(true);
+    });
+  });
+
+  describe('algorithm file parent', () => {
+    it('gets parent', async () => {
+      const file = Algorithmia.getClient(
+        process.env.ALGORITHMIA_DEFAULT_API_KEY
+      ).file('data://dherring/DalesNotSoFunTime2/NahDawg.txt');
```
Comment (bnlb) [inline test/Data.test.ts:44]: I'd still like this to be a dummy account that exists solely for the purpose of testing. Some username like `algorithmia-nodejs-e2e` or something obvious, then the API key used can come from that account.

File: test/Data.test.ts [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    it('gets for file', async () => {
+      const file = Algorithmia.getClient(
+        process.env.ALGORITHMIA_DEFAULT_API_KEY
+      ).file('data://dherring/DalesFunTime/NahDawg.txt');
+
+      expect(await file.get()).toBe('nah dawg');
+    });
+  });
+
+  describe('algorithm file head call', () => {
```
Comment (bnlb) [inline test/Data.test.ts:60]: Let's rename these test cases to reflect the method being tested, not the underlying HTTP method used.

```
describe('file.exists()', () => {
  it('should return true if a file exists', () => {
    // ...
  });
});

File: test/Data.test.ts [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+      ).dir('data://dherring/DalesNotSoFunTime3');
+
+      const directoryAlreadyExists = await dir.exists();
+      if (directoryAlreadyExists) {
+        await dir.delete(true);
+      }
+
+      await dir.post('DalesNotSoFunTime3');
+
+      expect(await dir.exists()).toBe(true);
```
Comment (bnlb) [inline test/Data.test.ts:107]: This test is failing for me.

File: test/Data.test.ts [hunk 17]
```diff
@@ -0,0 +1,158 @@
+import { Algorithmia } from '../src/Algorithmia';
+
+describe('Localisation initialization', () => {
+  beforeEach(() => {
+    jest.resetModules();
+  });
+
+  describe('algorithm file put call', () => {
```
Comment (bnlb) [inline test/Data.test.ts:8]: Just adding a note that once all the missing API methods that currently exist are added, tests should be created for those.

**Review by bnlb — CHANGES_REQUESTED** [review]
Main blockers:

- ~src/index.ts should have a default export of `Algorithmia`~
- The `examples/` folder should be added again (and updated if necessary to reflect API changes)
- ~Verify that the readme examples are all accurate after update~
- Add a CHANGELOG describing breaking changes (dropping support for callbacks in favor of promises, updating node version)
- Add missing methods that currently exist (`putFile`)
- ~Update tests with test cases once missing methods are added (`putFile`)
- ~Use a dummy `algorithmia-nodejs-e2e` user for tests instead of a personal account.~

File: test/Data.test.ts [hunk 18]
```diff
@@ -57,8 +123,38 @@ describe('Localisation initialization', () => {
...
+    });
+  });
+
+  describe('algorithm file getBinary call', () => {
+    it('gets file', async () => {
+      const file = Algorithmia.getClient(
+        process.env.ALGORITHMIA_DEFAULT_API_KEY
+      ).file('data://dherring/DalesFunTime/NahDawg.txt');
+
+      expect(await file.getBinary()).toBe('nah dawg');
```
Comment (bnlb) [inline test/Data.test.ts:152]: Since the function is `getBinary`, I'd expect something [like this](https://github.com/algorithmiaio/algorithmia-nodejs/blob/master/examples/img-read.js).

**Comment by bnlb** [comment]
@DaleHerring I've updated the blockers comment to reflect what's left.

**Review by bnlb — APPROVED** [review]
